docs/v1/P4Runtime-Spec.mdk

```diff
@@ -2320,6 +2325,116 @@ following semantics.
+implemented with an action selector, in one shot. This means that a
+table entry, an action profile group, and a set of action profile
+members can programmed with a single update message. Using one shots
+has the advanatage that the controller does not need to keep track of
+group ids and member ids.
+
+One shots are programmed by choosing the `ActionProfileActions`
+message as the `TableAction`. The `ActionProfileActions` message
+consists of a set of `ActionProfileAction` messages, which in turn
+have the following fields:
+
+* `action` is one of the actions specified by the table that is being
+  programmed.
+
+* `weight` specifying the probability of the action's selection at
+  runtime.
+
+* `watch` is the controller defined 32-bit port number that the
+  action's liveness depends on. At runtime, the action must be
+  excluded from selection if the this watch port is down.
+
+Semantically, one shots are equivalent to programming the table entry,
+group, and members individually. The necessary group id and member ids
+are bound to unused ids.
+
+For example, consider again the action selector defined in the
+beginning of this section. This table could be programmed with the
+following oneshot update:
+
+~ Begin Prototext
+table_entry {
+  table_id: 1
+  match { /* lpm match */ }
+  action {
+    action_profile_actions {
+      action_profile_action {
+        action { /* set nexthop 1 */ }
+        weight: 1
+        watch: 1
+      }
+      action_profile_action {
+        action { /* set nexthop 2 */ }
+        weight: 2
+        watch: 2
+      }
+      action_profile_action {
+        action { /* set nexthop 3 */ }
+        weight: 3
+        watch: 3
+      }
+    }
+  }
+}
+~ End Prototext
+
+Which would be equivalent to the following updates, where `GROUP_ID`,
+`MEMBER_ID_1`, `MEMBER_ID_2`, and `MEMBER_ID_3` are unused ids:
+
+~ Begin Prototext
+table_entry {
+  table_id: 1
+  match { /* lpm match */ }
+  action { action_profile_group_id: GROUP_ID }
+}
+action_profile_group {
+  action_profile_id: 1
+  group_id: GROUP_ID
+  members {
+    member_id: MEMBER_ID_1
+    weight: 1
+    watch: 1
+  }
+  members {
+    member_id: MEMBER_ID_2
+    weight: 2
+    watch: 2
+  }
+  members {
+    member_id: MEMBER_ID_3
+    weight: 3
+    watch: 3
+  }
+}
+action_profile_member {
+  action_profile_id: 1
+  member_id: MEMBER_ID_1
+  action { /* set nexthop 1 */ }
+}
+action_profile_member {
+  action_profile_id: 1
+  member_id: MEMBER_ID_2
+  action { /* set nexthop 2 */ }
+}
+action_profile_member {
+  action_profile_id: 1
+  member_id: MEMBER_ID_3
+  action {  /* set nexthop 3 */ }
+}
+~ End Prototext
+
+All the tables associated with an action selector may either be
+programmed exclusively with one shots, or exclusively with
+`ActionProfileMember` and `ActionProfileGroup` messages. Programming
+some entries with one shots, and other entries with
+`ActionProfileMember` and `ActionProfileGroup` messages is not
+allowed.
```

Comment:
Is a server free to support only the one-shot method?

Reply:
I would say yes, the spec should say that
 * a server MUST support the one shot method
 * it is recommended for a server to support both methods
I think the "MUST" for the first bullet point is justified by the fact that if you support the id-based method there is no reason not to be able to support the one shot method.

Reply:
Done.